src/query_builder.rs

```diff
@@ -0,0 +1,737 @@
+        Ok(tuple)
+    }
+
+    pub fn execute_rel(mut self) -> Result<RelResult> {
+        let results = self.execute()?;
+        let find_spec = results.spec.columns().map(|e| format!("{}", e)).collect();
+        let tuple = RelResult::new(find_spec, results.into_rel()?);
+        Ok(tuple)
+    }
+}
+
+pub trait VarResultsRow {
+    fn var_as_entid(&self, var: &str) -> Option<Entid>;
+    fn var_as_kw(&self, var: &str) -> Option<Rc<NamespacedKeyword>>;
+    fn var_as_boolean(&self, var: &str) -> Option<bool>;
+    fn var_as_long(&self, var: &str) -> Option<i64>;
+    fn var_as_double(&self, var: &str) -> Option<f64>;
+    fn var_as_instant(&self, var: &str) -> Option<DateTime<Utc>>;
+    fn var_as_timestamp(&self, var: &str) -> Option<i64>;
+    fn var_as_string(&self, var: &str) -> Option<Rc<String>>;
+    fn var_as_uuid(&self, var: &str) -> Option<Uuid>;
+    fn var_as_uuid_str(&self, var: &str) -> Option<String>;
+}
+
+pub trait ColResultsRow {
+    fn col_as_entid(&self, col: usize) -> Option<Entid>;
+    fn col_as_kw(&self, col: usize) -> Option<Rc<NamespacedKeyword>>;
+    fn col_as_boolean(&self, col: usize) -> Option<bool>;
+    fn col_as_long(&self, col: usize) -> Option<i64>;
+    fn col_as_double(&self, col: usize) -> Option<f64>;
+    fn col_as_instant(&self, col: usize) -> Option<DateTime<Utc>>;
+    fn col_as_timestamp(&self, col: usize) -> Option<i64>;
+    fn col_as_string(&self, col: usize) -> Option<Rc<String>>;
+    fn col_as_uuid(&self, col: usize) -> Option<Uuid>;
+    fn col_as_uuid_string(&self, col: usize) -> Option<String>;
+}
+
+pub struct ScalarResult(Option<TypedValue>);
+
+impl TypedValueConverter for ScalarResult {
+    fn as_entid(mut self) -> Option<Entid> {
+        self.0.take().map_or(None, |t| t.as_entid())
+    }
+
+    fn as_kw(mut self) -> Option<Rc<NamespacedKeyword>> {
+        self.0.take().map_or(None, |t| t.as_kw())
+    }
+
+    fn as_boolean(mut self) -> Option<bool> {
+        self.0.take().map_or(None, |t| t.as_boolean())
+    }
+
+    fn as_long(mut self) -> Option<i64> {
+        self.0.take().map_or(None, |t| t.as_long())
+    }
+
+    fn as_double(mut self) -> Option<f64> {
+        self.0.take().map_or(None, |t| t.as_double())
+    }
+
+    fn as_instant(mut self) -> Option<DateTime<Utc>> {
+        self.0.take().map_or(None, |t| t.as_instant())
+    }
+
+    fn as_timestamp(mut self) -> Option<i64> {
+        self.0.take().map_or(None, |t| t.as_timestamp())
+    }
+
+    fn as_string(mut self) -> Option<Rc<String>> {
+        self.0.take().map_or(None, |t| t.as_string())
+    }
+
+    fn as_uuid(mut self) -> Option<Uuid> {
+        self.0.take().map_or(None, |t| t.as_uuid())
+    }
+
+    fn as_uuid_string(mut self) -> Option<String> {
+        self.0.take().map_or(None, |t| t.as_uuid_string())
+    }
+}
+
+pub struct CollResult(pub Vec<TypedValue>);
+
+impl CollResult {
+    fn get_row(&self, row: usize) -> Option<TypedValue> {
+        self.0.get(row).map(|v| v.to_owned())
```

Comment:
This is a bad pattern to establish. Imagine a consumer that does:

```
let results = query.execute_coll();
let count = result.len();
for let i = 0..count {
    let s = results.col_as_uuid_string(i).unwrap();
    println!(s);
}
```

This `get_row` call will clone the UUID in order to copy the value out of the collection, and then `col_as_uuid_string` will allocate a new string.

We really want this query interface to be efficient, and that means not cloning unnecessarily. I think that means making this result interface simpler.